File: breadth-first-search/breadth_first_search.go
package main
import "fmt"

func bfs(graph map[int][]int, start int) {
    visited := make(map[int]bool)
    queue := []int{start}
    visited[start] = true
    
    for len(queue) > 0 {
        node := queue[0]
        queue = queue[1:]
        fmt.Print(node, " ")
        
        for _, neighbor := range graph[node] {
            if !visited[neighbor] {
                visited[neighbor] = true
                queue = append(queue, neighbor)
            }
        }
    }
}

func main() {
    graph := map[int][]int{
        0: {1, 2},
        1: {2},
        2: {0, 3},
        3: {3},
    }
    bfs(graph, 2)
}

File: breadth-first-search/breadth_first_search.cs
using System;
using System.Collections.Generic;

class BFS {
    static void Bfs(Dictionary<int, List<int>> graph, int start) {
        var visited = new HashSet<int>();
        var queue = new Queue<int>();
        
        visited.Add(start);
        queue.Enqueue(start);
        
        while (queue.Count > 0) {
            int node = queue.Dequeue();
            Console.Write(node + " ");
            
            if (graph.ContainsKey(node)) {
                foreach (int neighbor in graph[node]) {
                    if (!visited.Contains(neighbor)) {
                        visited.Add(neighbor);
                        queue.Enqueue(neighbor);
                    }
                }
            }
        }
    }
    
    static void Main() {
        var graph = new Dictionary<int, List<int>> {
            {0, new List<int> {1, 2}},
            {1, new List<int> {2}},
            {2, new List<int> {0, 3}},
            {3, new List<int> {3}}
        };
        Bfs(graph, 2);
    }
}